File: backend/models/advice.py
"""AI advice sessions and items."""
from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from core.db import Base

DISCLAIMER = "Это аналитическая информация, не финансовая рекомендация."


class AdviceSession(Base):
    __tablename__ = "advice_sessions"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), index=True
    )
    snapshot_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("snapshots.id"), nullable=True
    )
    generated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    model_used: Mapped[str | None] = mapped_column(String(50), nullable=True)
    tokens_used: Mapped[int | None] = mapped_column(Integer, nullable=True)
    advice_count: Mapped[int | None] = mapped_column(Integer, nullable=True)

    items: Mapped[list["AdviceItem"]] = relationship(
        back_populates="session", cascade="all, delete-orphan"
    )


class AdviceItem(Base):
    __tablename__ = "advice_items"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    session_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("advice_sessions.id", ondelete="CASCADE"), index=True
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), index=True
    )
    title: Mapped[str] = mapped_column(Text)
    category: Mapped[str | None] = mapped_column(String(30), nullable=True)
    body: Mapped[str] = mapped_column(Text)
    relevance: Mapped[str | None] = mapped_column(Text, nullable=True)
    disclaimer: Mapped[str] = mapped_column(Text, default=DISCLAIMER)
    is_read: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )

    session: Mapped["AdviceSession"] = relationship(back_populates="items")

File: frontend/src/services/advice.ts
import { api } from "./api";

export interface AdviceItem {
  id: string;
  title: string;
  category: string | null;
  body: string;
  relevance: string | null;
  disclaimer: string;
  is_read: boolean;
  created_at: string | null;
}

export interface AdviceSession {
  session_id: string;
  snapshot_id: string | null;
  generated_at: string | null;
  advice_count: number;
  items: AdviceItem[];
}

export async function generateAdvice(): Promise<AdviceSession> {
  const { data } = await api.post<AdviceSession>("/advice/generate");
  return data;
}

export async function getLatestAdvice(): Promise<AdviceSession | null> {
  const { data } = await api.get<AdviceSession | null>("/advice/latest");
  return data;
}

export async function markAdviceRead(itemId: string): Promise<void> {
  await api.post(`/advice/${itemId}/read`);
}
